pkgs/cli_config/lib/src/config.dart

```diff
@@ -0,0 +1,447 @@
+    _throwIfNull(key, value);
+    return value!;
+  }
+
+  /// Lookup an optional path in this config.
+  ///
+  /// First tries CLI argument defines, then environment variables, and
+  /// finally the config file.
+  ///
+  /// Throws if one of the configs does not contain the expected value type.
+  ///
+  /// If [resolveFileUri], resolves the paths in config file relative to the
+  /// config file.
+  ///
+  /// If [mustExist], throws if the path doesn't resolve to a file or directory
+  /// on the file system.
+  Uri? getOptionalPath(
+    String key, {
+    bool resolveFileUri = true,
+    bool mustExist = false,
+  }) {
+    for (final provider in _providers) {
+      final path = provider.getOptionalString(key);
+      if (path != null) {
+        final value = _pathToUri(
+          path,
+          resolveUri: resolveFileUri && provider == _fileProvider,
+          baseUri: provider.baseUri,
+        );
+        if (mustExist) {
+          _throwIfNotExists(key, value);
+        }
+        return value;
+      }
+    }
+    return null;
+  }
+
+  Uri _pathToUri(
+    String path, {
+    required bool resolveUri,
+    required Uri? baseUri,
+  }) {
+    if (resolveUri && baseUri != null) {
+      return baseUri.resolve(path);
+    }
+    return Provider.fileSystemPathToUri(path);
+  }
+
+  /// Lookup a list of paths in this config.
+  ///
+  /// If [combineAllConfigs] combines results from cli, environment, and
+  /// config file. Otherwise, precedence rules apply.
+  ///
+  /// If provided, [splitCliPattern] splits cli defines.
+  ///
+  /// If provided, [splitEnvironmentPattern] splits environment values.
+  ///
+  /// If [resolveFileUri], resolves the paths in config file relative to the
+  /// config file.
+  List<Uri>? getOptionalPathList(
+    String key, {
+    bool combineAllConfigs = true,
+    String? splitCliPattern,
+    String? splitEnvironmentPattern,
+    bool resolveFileUri = true,
+  }) {
+    List<Uri>? result;
+    for (final entry in {
+      _cliProvider: splitCliPattern,
+      _environmentProvider: splitEnvironmentPattern,
+      _fileProvider: null
+    }.entries) {
+      final provider = entry.key;
+      final splitPattern = entry.value;
+      final paths = provider.getOptionalStringList(
+        key,
+        splitPattern: splitPattern,
+      );
+      if (paths != null) {
+        final value = [
+          for (final path in paths)
+            _pathToUri(
+              path,
+              resolveUri: resolveFileUri && provider == _fileProvider,
+              baseUri: provider.baseUri,
+            )
+        ];
+        if (combineAllConfigs) {
+          (result ??= []).addAll(value);
+        } else {
+          return value;
+        }
+      }
+    }
+    return result;
+  }
+
+  /// Lookup a [T] in the config file.
+  ///
+  /// Only available for the configuration file, cannot be overwritten with
+  /// commandline defines or environment variables.
```

Comment:
> That is, have a general API, and have implementations which can only provide values for some of the features of that API, rather than having to define operations as only working on one version or the other.

Making it `optionalValueOf` and returning nothing if the source can't provide it sounds like a good idea.

> If every provider provided this method, but the other sources can only produce String values, it would open you up to having more sources in the future.

Maybe in the future we should actually support defining more complicated data types from env-variables or cl-defines. https://helm.sh/docs/intro/using_helm/ supports doing `--set foo[bar]=baz --set foo[baf]=ban` to create a map from CLI.

We kind of already have this with the 3 different `Provider`s. Which could be called `Source`s. Or in your terminology `Data`. I was thinking to not provide these `Source`s in the public API for now. Maybe we can expose them later if there is a need for it.